### PR TITLE
Bug 1741288 - Support multiple IP addresses for GCLBs

### PR DESCRIPTION
**Pull request by ANich**
Follow up to https://github.com/mozilla/gcp-ingestion/pull/1815/files#r703625278

- Updated the `ParseProxy` transform to filter from a CSV string for the static GCLB IPs instead of a single value.  
- Updated the [default IPs](https://bugzilla.mozilla.org/show_bug.cgi?id=1741288#c1).

### CODE REVIEW

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java [hunk 1]
```diff
@@ -14,11 +14,12 @@
  */
 public interface DecoderOptions extends SinkOptions, PipelineOptions {
 
-  @Description("Static IP associated with load balancer; will be ignored for IP lookups")
-  @Default.String("35.227.207.240")
-  String getProxyIpAddress();
+  @Description("Comma separated list of static IPs associated with load balancer; "
```
Comment (whd) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java:17]:
```suggestion
  @Description("Comma-separated list of static IPs associated with load balancers; "
```

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java [hunk 2]
```diff
@@ -14,11 +14,12 @@
...
+      + "will be ignored for IP lookups")
+  @Default.String("35.227.207.240,34.120.208.123")
+  String getProxyIpAddressCsv();
```
Comment (whd) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java:20]: I recommend `proxyIps` per https://github.com/mozilla-services/cloudops-infra/pull/3582/files#diff-8c0bf74f03d3c2a36089bf3c606bf05cbc746f13f5ad989f72341f28dfa91915R48 or perhaps `proxyIpAddresses`. We have a couple of CSV parameters and none are explicitly named such, so I would want to stick to avoid suffixing with `CSV` here.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java [hunk 3]
```diff
@@ -48,14 +48,16 @@ public PubsubMessage apply(PubsubMessage message) {
...
       // we will need to remove; see
       // https://bugzilla.mozilla.org/show_bug.cgi?id=1729069#c8
-      if (xff != null && xff.contains(proxyIpAddress)) {
-        xff = Arrays.stream(xff.split("\\s*,\\s*")) //
-            .filter(ip -> !proxyIpAddress.equals(ip)) //
-            .collect(Collectors.joining(","));
-        countProxyIpAddress.inc();
+      for (String proxyIpAddress : proxyIpAddressCsv.split(",")) {
+        if (xff != null && xff.contains(proxyIpAddress)) {
+          xff = Arrays.stream(xff.split("\\s*,\\s*")) //
```
Comment (whd) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:56]: I see elsewhere in this repo (ctxsvc-reporter job) the pattern of `.filter(StringUtils::isNotBlank)` instead of splitting with spaces as is done here. Perhaps this should be treated the same way.

Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:56]: This particular line is unchanged from previous implementation (just it's now nested in a for loop). If we ended up with a blank entry here, I think it would mean that the input `xff` included a trailing comma, which would be very unexpected and which might cause additional debugging confusion if we filter it out implicitly here.

Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:56]: But it would be good to to add `filter(StringUtils::isNotBlank)` when we do `proxyIpAddressCsv.split(",")` above.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java [hunk 4]
```diff
@@ -48,14 +48,16 @@ public PubsubMessage apply(PubsubMessage message) {
...
-      // If the configured proxyIpAddress is present, then there's a GCP load balancer IP
+      // If any of the configured proxyIpAddresses is present, then there's a GCP load balancer IP
       // we will need to remove; see
       // https://bugzilla.mozilla.org/show_bug.cgi?id=1729069#c8
-      if (xff != null && xff.contains(proxyIpAddress)) {
-        xff = Arrays.stream(xff.split("\\s*,\\s*")) //
-            .filter(ip -> !proxyIpAddress.equals(ip)) //
-            .collect(Collectors.joining(","));
-        countProxyIpAddress.inc();
+      for (String proxyIpAddress : proxyIpAddressCsv.split(",")) {
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:54]: I think we should parse this parameter into a `Set<String>` within the constructor. Then we can check `ip -> !proxyIpAddresses.contains(ip)` and avoid having the outer for loop.

We don't do that sort of thing in existing code because we only recently transitioned from "classic templates" to Flex Templates. With classic templates, this would have been a runtime parameter that would not have been defined when the constructor was called. But now all pipeline options are fully available when the code is compiled.

**Review by jklukas — CHANGES_REQUESTED** [review]
Nicely done, @ANich. This looks functionally correct to me.

I left some suggestions for additional cleanup. But renaming the option to one of @whd's suggestions is the only blocker here.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java (hunk 4 above)
Comment (ANich) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:54]: How would `countProxyIpAddress.inc()` work in this case? Is it not counting the number of proxy ips that get matched? (Previously there would only have been one)

Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java:54]: Good question. We'd only ever expect one proxy IP to be present in `xff`. I think it's fine to just increment the counter once based on `proxyIpAddresses.contains(ip)` being true.

Comment (ANich) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParseProxy.java]: I have a version that removes the outer loop. Something like:
```
private final Set<String> proxyIps;
...
if (xff != null) { // && at least one ip in the xff appears in the proxyIps {
        xff = Arrays.stream(xff.split("\\s*,\\s*")) //
            .filter(StringUtils::isNotBlank) //
            .filter(ip -> !proxyIps.contains(ip)) //
            .collect(Collectors.joining(","));
        countProxyIpAddress.inc();
 }
```
The extra clause would be something like "if the set of xff ips intersects the set of `proxyIps`" since each `ip` is only in scope in the closures unless I'm missing something... Is that what you had in mind?

**Review by jklukas — APPROVED** [review]
This looks functionally correct and has a good interface now. @whd can feel free to merge

**Comment by whd** [comment]
tl;dr In future I will update the PR title with `[do not merge]` or similar for PRs I expect to merge myself.

Per https://bugzilla.mozilla.org/show_bug.cgi?id=1741288#c1 I was expecting to be responsible for this merge, since it involves coordinated deployment with another PR (https://github.com/mozilla-services/cloudops-infra/pull/3582). In practice it was fine to merge this for two reasons:

1. The default value for this parameter is the value that prod uses
1. Beam code changes only deploy to stage automatically (#1458)

This ended up deploying to stage and "soft failing" in that the decoder jobs successfully started but after three minutes failed, see https://console.cloud.google.com/dataflow/jobs/us-west1/2021-12-02_08_46_53-923418715400192148?project=moz-fx-data-beam-nonprod-b5f4.

It is interesting to observe this failure mode (specific to flex templates I think): `2021-12-02 16:50:26.202 GMTError occurred in the launcher container: Template launch failed. See console logs.`. Switching to INFO I see
```2021-12-02 16:46:54.700 GMTConsole log from launcher will be available at gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/console_logs.```

And the log here shows what we'd expect (`java.lang.IllegalArgumentException: Class interface com.mozilla.telemetry.decoder.DecoderOptions missing a property named 'proxyIpAddress'`, since we're still using the old value in ops logic):
```
gsutil cat gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/console_logs
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.403536","line":"java_template_launcher.go:40","message":"Started template launcher."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.404896","line":"java_template_launcher.go:44","message":"Initialize Java template."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.406830","line":"java_template.go:100","message":"Validating metadata template-container-args: {\"consoleLogsLocation\":\"gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/console_logs\",\"environment\":{\"region\":\"us-west1\",\"serviceAccountEmail\":\"beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com\",\"stagingLocation\":\"gs://dataflow-staging-us-west1-151430467146/staging\",\"tempLocation\":\"gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/\"},\"jobId\":\"2021-12-02_08_46_53-923418715400192148\",\"jobName\":\"telemetry-raw-decoder-1c9d55c1-1\",\"jobObjectLocation\":\"gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/job_object\",\"operationResultLocation\":\"gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/operation_result\",\"parameters\":{\"autoscalingAlgorithm\":\"THROUGHPUT_BASED\",\"enableStreamingEngine\":\"false\",\"errorOutput\":\"projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-error\",\"errorOutputNumShards\":\"60\",\"errorOutputType\":\"pubsub\",\"geoCityDatabase\":\"gs://moz-fx-data-prod-geoip/GeoIP2-City/20211203/GeoIP2-City.mmdb\",\"geoCityFilter\":\"gs://moz-fx-data-prod-dataflow-templates/cities15000.txt\",\"geoIspDatabase\":\"gs://moz-fx-data-prod-geoip/GeoIP2-ISP/20211203/GeoIP2-ISP.mmdb\",\"input\":\"projects/moz-fx-data-beam-nonprod-b5f4/subscriptions/telemetry-raw.decoder\",\"inputFileFormat\":\"json\",\"inputType\":\"pubsub\",\"maxNumWorkers\":\"6\",\"numWorkers\":\"1\",\"output\":\"projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-decoded\",\"outputFileFormat\":\"json\",\"outputNumShards\":\"200\",\"outputType\":\"pubsub\",\"proxyIpAddress\":\"34.98.105.174\",\"pubsubIdAttribute\":\"uri\",\"schemasLocation\":\"gs://moz-fx-data-prod-dataflow/schemas/202112020304_ac48d9a4.tar.gz\",\"serviceAccount\":\"beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com\",\"stagingLocation\":\"gs://dataflow-staging-us-west1-151430467146/staging\",\"tempLocation\":\"gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/\",\"usePublicIps\":\"false\",\"windowDuration\":\"1m\"},\"projectId\":\"moz-fx-data-beam-nonprod-b5f4\"}"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407345","line":"java_template.go:109","message":"Extracting operation result location."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407372","line":"java_template.go:117","message":"Operation result location: gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/operation_result"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407391","line":"java_template.go:120","message":"Extracting console log location."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407403","line":"java_template.go:128","message":"Console logs location: gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/console_logs"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407421","line":"java_template.go:131","message":"Extracting and validating Java command specs."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407691","line":"java_template.go:139","message":"Generating launch args."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407769","line":"java_args.go:136","message":"Overriding stagingLocation with value: gs://dataflow-staging-us-west1-151430467146/staging (previous value: gs://dataflow-staging-us-west1-151430467146/staging)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407795","line":"java_args.go:136","message":"Overriding tempLocation with value: gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/ (previous value: gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407816","line":"java_args.go:136","message":"Overriding serviceAccount with value: beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com (previous value: beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407858","line":"launch.go:47","message":"Validating ExpectedFeatures."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407880","line":"launch.go:72","message":"Launching Java template."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407909","line":"java_template.go:63","message":"Using launch args: [-cp /template/ingestion-beam-0.1-SNAPSHOT.jar com.mozilla.telemetry.Decoder --input=projects/moz-fx-data-beam-nonprod-b5f4/subscriptions/telemetry-raw.decoder --tempLocation=gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/ --serviceAccount=beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com --maxNumWorkers=6 --windowDuration=1m --outputNumShards=200 --usePublicIps=false --output=projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-decoded --proxyIpAddress=34.98.105.174 --project=moz-fx-data-beam-nonprod-b5f4 --numWorkers=1 --errorOutputNumShards=60 --geoIspDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-ISP/20211203/GeoIP2-ISP.mmdb --jobName=telemetry-raw-decoder-1c9d55c1-1 --errorOutputType=pubsub --runner=DataflowRunner --stagingLocation=gs://dataflow-staging-us-west1-151430467146/staging --inputFileFormat=json --errorOutput=projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-error --geoCityDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-City/20211203/GeoIP2-City.mmdb --geoCityFilter=gs://moz-fx-data-prod-dataflow-templates/cities15000.txt --templateLocation=gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/job_object --region=us-west1 --outputType=pubsub --enableStreamingEngine=false --schemasLocation=gs://moz-fx-data-prod-dataflow/schemas/202112020304_ac48d9a4.tar.gz --inputType=pubsub --outputFileFormat=json --pubsubIdAttribute=uri --autoscalingAlgorithm=THROUGHPUT_BASED]"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:49.407994","line":"exec.go:37","message":"Executing: java -cp /template/ingestion-beam-0.1-SNAPSHOT.jar com.mozilla.telemetry.Decoder --input=projects/moz-fx-data-beam-nonprod-b5f4/subscriptions/telemetry-raw.decoder --tempLocation=gs://moz-fx-data-beam-nonprod-b5f4-staging/telemetry-raw-decoder-1c9d55c1-1/ --serviceAccount=beam-stage-decoder@moz-fx-data-beam-nonprod-b5f4.iam.gserviceaccount.com --maxNumWorkers=6 --windowDuration=1m --outputNumShards=200 --usePublicIps=false --output=projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-decoded --proxyIpAddress=34.98.105.174 --project=moz-fx-data-beam-nonprod-b5f4 --numWorkers=1 --errorOutputNumShards=60 --geoIspDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-ISP/20211203/GeoIP2-ISP.mmdb --jobName=telemetry-raw-decoder-1c9d55c1-1 --errorOutputType=pubsub --runner=DataflowRunner --stagingLocation=gs://dataflow-staging-us-west1-151430467146/staging --inputFileFormat=json --errorOutput=projects/moz-fx-data-shar-nonprod-efed/topics/telemetry-error --geoCityDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-City/20211203/GeoIP2-City.mmdb --geoCityFilter=gs://moz-fx-data-prod-dataflow-templates/cities15000.txt --templateLocation=gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/job_object --region=us-west1 --outputType=pubsub --enableStreamingEngine=false --schemasLocation=gs://moz-fx-data-prod-dataflow/schemas/202112020304_ac48d9a4.tar.gz --inputType=pubsub --outputFileFormat=json --pubsubIdAttribute=uri --autoscalingAlgorithm=THROUGHPUT_BASED"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.958749","line":"exec.go:64","message":"Exception in thread \"main\" "}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.959128","line":"exec.go:64","message":"java.lang.IllegalArgumentException: Class interface com.mozilla.telemetry.decoder.DecoderOptions missing a property named 'proxyIpAddress'."}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.959327","line":"exec.go:64","message":"\tat org.apache.beam.sdk.options.PipelineOptionsFactory.parseObjects(PipelineOptionsFactory.java:1874)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.959888","line":"exec.go:64","message":"\tat org.apache.beam.sdk.options.PipelineOptionsFactory.access$400(PipelineOptionsFactory.java:144)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.960021","line":"exec.go:64","message":"\tat org.apache.beam.sdk.options.PipelineOptionsFactory$Builder.as(PipelineOptionsFactory.java:327)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.960119","line":"exec.go:64","message":"\tat com.mozilla.telemetry.Decoder.run(Decoder.java:46)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.960211","line":"exec.go:64","message":"\tat com.mozilla.telemetry.Decoder.main(Decoder.java:35)"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.989899","line":"exec.go:50","message":"java failed with exit status 1"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.989959","line":"launch.go:77","message":"Template launch failed: exit status 1"}
{"container_id":"51f0d060480f6025927386e0626b0dc19f3a78e8bd6e0fafc0266f615f606426","severity":"INFO","time":"2021/12/02 16:47:52.989983","line":"launch.go:99","message":"Uploading console logs to gcs location: gs://dataflow-staging-us-west1-151430467146/staging/template_launches/2021-12-02_08_46_53-923418715400192148/console_logs"}
```

This ended up paging on stage decoder latency, but I had already noticed something was going on by the time that happened because @jklukas had pinged me. It might be better to implement some monitoring on dataflow job status instead to catch this more quickly. I will be more explicit in comments in the future when I review changes that I intend to merge myself.

**Comment by jklukas** [comment]
Note that I specifically gave @ANich the go-ahead to merge this, so definitely my fault for missing this possibility. It totally makes sense that changing the options interface would have downstream effects.

**Comment by whd** [comment]
I also mentioned it in both comments of https://bugzilla.mozilla.org/show_bug.cgi?id=1741288 (embedded with other things), but it's a much clearer a signal when it's in the PR title, so I'll make sure to do that in the future.

**Comment by ANich** [comment]
Ahh yes.. I definitely saw this:
> These PRs need to coordinated so I will be responsible for final merge and deploy, probably next week.

last week but missed it on the second pass after squashing and rebasing my bad!
